src/rust/lib/system/web/src/platform.rs

```diff
@@ -0,0 +1,42 @@
+//! This module provides helpers for platform specific logic.
+
+use super::window;
+
+/// This enumeration lists all the supported platforms.
+#[derive(Debug,Clone,Copy)]
+pub enum Platform {
+    Linux,
+    Android,
+    Windows,
+    MacOS,
+    IOS,
+    Unknown
+}
+
+impl Platform {
+    /// Queries which platform we are on.
+    pub fn query() -> Self {
+        let window = window();
+        let navigator = window.navigator();
+        let platform  = navigator.platform().unwrap_or_else(|_| "Unknown".into());
+        let agent     = navigator.user_agent().unwrap_or_else(|_| "Unknown".into());
+
+        const MACOS_PLATFORMS   : [&str; 4] = ["Macintosh", "MacIntel", "MacPPC", "Mac68K"];
+        const WINDOWS_PLATFORMS : [&str; 4] = ["Win32", "Win64", "Windows", "WinCE"];
+        const IOS_PLATFORMS     : [&str; 3] = ["iPhone", "iPad", "iPod"];
+
+        if MACOS_PLATFORMS.iter().any(|item| **item == platform) {
+            Platform::MacOS
+        } else if IOS_PLATFORMS.iter().any(|item| **item == platform) {
+            Platform::IOS
+        } else if WINDOWS_PLATFORMS.iter().any(|item| **item == platform) {
+            Platform::Windows
+        } else if agent.find("Android").is_some() {
+            Platform::Android
+        } else if platform.find("Linux").is_some() {
+            Platform::Linux
```

Comment:
Let's keep all the "platform constants" in one place (so put the const (LINUX|ANDROID)_PLATFORM : &str = next to rest of platforms list.

Reply:
We can't generalize this logic because `Android` needs to be searched in `agent` and `Linux` is searched in `platform`.

Reply:
Do we always have either `navigator.platform()` or `navigator.user_agent()`? Are both valid on Android / Linux or they are used there interchangeable. Im asking cause I also think this code should be more structured (as Adam suggested) and I have idea how to do it but I need the above info :)

Reply:
@wdanilo yes, we need both `platform` and `user_agent`. Please check the following links:

https://stackoverflow.com/questions/19877924/what-is-the-list-of-possible-values-for-navigator-platform-as-of-today

https://developer.chrome.com/multidevice/user-agent

Reply:
Oh god. Based on the linked answer Im wondering why we just don't do everything in unified way - search if there was `win`, `mac`, `linux`, `ipad`, `iphone` string inside the `platform` string (I used lowercase letters intentionally - we want to catch when browser decides to return "windows" instead of "Windows". All these things are not specified anywhere so we will have hard time tracking down what happened when a browser changes its way of reporting it. so lets make it as generic as possible. What do you think about it?